homeassistant/components/mqtt/mixins.py

```diff
@@ -1113,14 +1119,52 @@ async def async_publish(
     def config_schema() -> vol.Schema:
         """Return the config schema."""
 
+    def _set_entity_name(self, config: ConfigType) -> None:
+        """Help setting the entity name if needed."""
+        entity_name: str | None | UndefinedType = config.get(CONF_NAME, UNDEFINED)
+        # Only set _attr_name if it is needed
+        if entity_name is not UNDEFINED:
+            self._attr_name = entity_name
+        elif (
+            not (use_device_class_name := self._default_to_device_class_name())
+            and self._default_name is not UNDEFINED
+        ):
+            # Assign the default name
+            self._attr_name = self._default_name
+        elif use_device_class_name:
+            # Follow name of device class
+            self._attr_has_entity_name = True
+            if CONF_DEVICE in config and CONF_NAME not in config[CONF_DEVICE]:
+                _LOGGER.warning(
+                    "No device name set in config: %s,"
+                    "'has_entity_name' was set to True, please set a device name, see also "
+                    "https://developers.home-assistant.io/docs/core/entity/#entity-naming",
+                    config,
+                )
+            return
+        if (device_config := CONF_DEVICE in config) and CONF_NAME in config[
+            CONF_DEVICE
+        ]:
+            # Only set has_entity_name if a valid name is set
+            self._attr_has_entity_name = True
+        elif device_config:
+            self._attr_has_entity_name = False
+            _LOGGER.warning(
+                "No device name set in config: %s,"
+                "'has_entity_name' was set to False, this is not recommended, see also "
+                "https://developers.home-assistant.io/docs/core/entity/#entity-naming",
+                config,
+            )
+
```

Comment:
This is quite unreadable, I'd suggest to simplify it
```suggestion
        if CONF_DEVICE in config:
            self._attr_has_entity_name = device_named = CONF_NAME in config
            if not device_named:
                _LOGGER.warning(
                    "No device name set in config: %s,"
                    "'has_entity_name' was set to False, this is not recommended, see also "
                    "https://developers.home-assistant.io/docs/core/entity/#entity-naming",
                    config,
                )
```

Reply:
Why is not not the same logic?